File: q10-q19/q16/q16.cpp
#include <algorithm>
#include <cmath>
class Solution {
public:
    int threeSumClosest(vector<int>& nums, int target) {
        int solution = 0;
        int current_diff = 0;
        int smallest_diff = INT_MAX;
        int current_sum = 0;
        std::sort(nums.begin(), nums.end());
        for (int i = 0; i < nums.size()-2; i++){
            int left = i + 1;
            int right = nums.size() - 1;
            while(left < right){
                current_sum = nums[i] + nums[right] + nums[left];
                if(abs(target - current_sum) < smallest_diff){
                    smallest_diff = abs(target - current_sum);
                    solution = current_sum;
                }
                if (current_sum < target)
                    left++;
                else if (current_sum > target)
                    right--;
                else
                    return target;
            }   
        }
        return solution;
    }
};

File: q1-q99/q10-q19/q16/q16.py
class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        solution = 0;
        current_sum = 0;
        current_diff = inf;
        nums.sort();
        for i in range(0, len(nums)-2):
            left = i + 1;
            right = len(nums) - 1;
            while left < right:
                current_sum = nums[i] + nums[left] + nums[right];
                if abs(target - current_sum) < current_diff:
                    current_diff = abs(target - current_sum);
                    solution = current_sum;
                if current_sum < target:
                    left += 1;
                elif current_sum > target:
                    right -= 1;
                else:
                    return target;
        return solution;
